crates/sui-core/src/execution_engine.rs

```diff
@@ -470,6 +535,121 @@ fn pay<S>(
     Ok(())
 }
 
+fn update_input_coins<S>(
+    temporary_store: &mut TemporaryStore<S>,
+    coin_objects: &mut Vec<Object>,
+    gas_coin: &Coin,
+    recipient: Option<SuiAddress>,
+) {
+    let mut gas_coin_obj = coin_objects.remove(0);
+    // unwrap is safe because we checked that it was a coin object above.
+    // update_contents_without_version_change b/c this is the gas coin,
+    // whose version will be bumped upon gas payment.
+    gas_coin_obj
+        .data
+        .try_as_move_mut()
+        .unwrap()
+        .update_contents_without_version_change(
+            bcs::to_bytes(gas_coin).expect("Coin serialization should not fail"),
+        );
+    if let Some(recipient) = recipient {
+        gas_coin_obj.transfer_without_version_change(recipient);
+    }
+    temporary_store.write_object(gas_coin_obj, WriteKind::Mutate);
+
+    for coin_object in coin_objects.iter() {
+        temporary_store.delete_object(&coin_object.id(), coin_object.version(), DeleteKind::Normal)
+    }
+}
+
+fn pay_sui<S>(
+    temporary_store: &mut TemporaryStore<S>,
+    coin_objects: &mut Vec<Object>,
+    recipients: Vec<SuiAddress>,
+    amounts: Vec<u64>,
+    tx_ctx: &mut TxContext,
+) -> Result<(), ExecutionError> {
+    let (mut coins, coin_type) = check_coins(coin_objects)?;
```

Comment:
I think it's cleaner if you could define `check_coins` to take an optional coin type for the check.